main.go

```diff
@@ -19,105 +17,122 @@ type FileInfo map[string]time.Time
 	flag.Parse()
 
-	watchedFiles = walk(rootPath)
-	scheduleScanAt(rootPath, interval)
-}
+	if *interval < 0 {
+		log.Fatal("negative interval:", *interval)
+	}
 
-func scheduleScanAt(rootPath *string, interval *int) {
-	runTicker := time.NewTicker(
-		time.Duration(time.Duration(*interval) * time.Second))
-	for {
-		select {
-		case <-runTicker.C:
-			scan(rootPath)
-		}
+	if err := os.Chdir(*rootPath); err != nil {
+		log.Fatal("Chdir:", err)
+	}
+
+	watchedFiles = walk(rootPath)
+	for range time.NewTicker(*interval).C {
+		scan(rootPath)
 	}
 }
 
 func scan(rootPath *string) {
+	changedDirs := make(map[string]bool, 0)
 	for filePath, mostRecentModTime := range walk(rootPath) {
-		if lastModTime, ok := watchedFiles[filePath]; ok {
-			if lastModTime != mostRecentModTime {
-				watchedFiles[filePath] = mostRecentModTime
-				runOrSkipTest(filePath)
-			}
-		} else {
+		lastModTime, ok := watchedFiles[filePath]
+		if !ok {
 			// files recently added
-			fileInfo, _ := os.Stat(filePath)
+			fileInfo, err := os.Stat(filePath)
+			if err != nil {
+				log.Print("Stat:", filePath, err)
+				continue
+			}
 			watchedFiles[filePath] = fileInfo.ModTime()
+			continue
 		}
+
+		// no changes
+		if lastModTime == mostRecentModTime {
+			continue
+		}
+
+		watchedFiles[filePath] = mostRecentModTime
+		if !needsTest(filePath) {
+			continue
+		}
+
+		pkgDir := path.Dir(filePath)
+		changedDirs[pkgDir] = true
+	}
+
+	if len(changedDirs) == 0 {
+		return
 	}
+
+	dedup := make([]string, 0)
+	for dir := range changedDirs {
+		dedup = append(dedup, dir)
+	}
+	test(dedup)
 }
 
 func walk(rootPath *string) FileInfo {
-	watchedFiles := FileInfo{}
-	err := filepath.Walk(*rootPath, visit(watchedFiles))
-	if err != nil {
-		panic(err)
+	wf := FileInfo{}
+	if err := filepath.Walk(*rootPath, visit(wf)); err != nil {
+		log.Fatal("Walk:", err)
 	}
-
-	return watchedFiles
+	return wf
 }
 
-func visit(watchedFiles FileInfo) filepath.WalkFunc {
+func visit(wf FileInfo) filepath.WalkFunc {
 	return func(path string, info os.FileInfo, err error) error {
 		if err != nil {
-			log.Fatal(err)
+			return err
 		}
 
 		if filepath.Ext(path) == ".go" {
-			watchedFiles[path] = info.ModTime()
+			wf[path] = info.ModTime()
 		}
 
 		return nil
 	}
 }
 
-func runOrSkipTest(filePath string) {
-	if isRegularFile(filePath) {
-		// run the respective test for go files
-		testFilePath := findTestFilePath(filePath)
-		if testFilePath != "" {
-			test(testFilePath)
-		}
-	} else {
-		// run tests directly
-		test(filePath)
+// needsTest returns if we need to run test for the changed path.
+func needsTest(filePath string) bool {
+	if isTestFile(filePath) {
+		// change happened on a test file, we need to test.
+		return true
 	}
-}
 
-func findTestFilePath(filePath string) string {
+	// changed file is not a test file, we return true only if there is a
+	// test file for the go file.
 	fileName := path.Base(filePath)
```

Comment:
I'd turn this logic into a new function like `func hasTestFile(filePath string) bool`. In this case, this function stays more readable like:

```
func needsTest(filePath string) bool {
	if isTestFile(filePath) || hasTestFile(filePath) {
		return true
	}

	return false
}
```